server/lib/two-factor-authentication/lib.ts

```diff
@@ -0,0 +1,168 @@
+
+const DEFAULT_TWO_FACTOR_AUTH_SESSION_DURATION = 600; // 10min
+
+type ValidateRequestOptions = {
+  // require user configured 2FA
+  requireTwoFactorAuthEnabled?: boolean;
+  // always ask for a token when using 2FA
+  alwaysAskForToken?: boolean;
+  // duration which we wont require a token after a successful use
+  sessionDuration?: number;
+  // identifier for the session
+  sessionKey?: (() => string) | string;
+};
+
+export enum TwoFactorMethod {
+  TOTP = 'totp',
+}
+
+export const TwoFactorAuthenticationHeader = 'x-two-factor-authentication';
+
+export const SupportedTwoFactorMethods = [TwoFactorMethod.TOTP];
+
+export type Token = {
+  type: TwoFactorMethod;
+  code: string;
+};
+
+export interface TwoFactorAuthProvider {
+  validateToken(user: typeof User, token: Token): Promise<void>;
+}
+
+export const providers: { [method in TwoFactorMethod]: TwoFactorAuthProvider } = {
+  [TwoFactorMethod.TOTP]: totp,
+};
+
+function getTwoFactorAuthTokenFromRequest(req: Request): Token {
+  const header = req.get(TwoFactorAuthenticationHeader);
+  if (!header) {
+    return null;
+  }
+
+  const parts = header.split(' ');
+  const type = parts[0] as TwoFactorMethod;
+  const code = parts[1];
+
+  return {
+    type,
+    code,
+  };
+}
+
+async function validateToken(user: typeof User, token: Token): Promise<void> {
+  if (!SupportedTwoFactorMethods.includes(token.type)) {
+    throw new Error(`Unsupported 2FA type ${token.type}`);
+  }
+
+  return providers[token.type].validateToken(user, token);
+}
+
+const DefaultValidateRequestOptions: ValidateRequestOptions = {
+  requireTwoFactorAuthEnabled: false,
+  alwaysAskForToken: false,
+  sessionDuration: DEFAULT_TWO_FACTOR_AUTH_SESSION_DURATION,
+};
+
+function getSessionKey(req: Request, options: ValidateRequestOptions) {
+  if (typeof options.sessionKey === 'function') {
+    return options.sessionKey();
+  } else if (typeof options.sessionKey === 'string') {
+    return options.sessionKey;
+  } else {
+    const userId = req?.remoteUser?.id;
+    const sessionId = req.jwtPayload?.sessionId;
+    return `${userId}_2fa_${sessionId}`;
+  }
+}
+
+async function hasValidTwoFactorSession(
+  req: Request,
+  options: ValidateRequestOptions = DefaultValidateRequestOptions,
+): Promise<boolean> {
+  const sessionKey = getSessionKey(req, options);
+
+  const twoFactorSession = await cache.get(sessionKey);
+
+  if (isNil(twoFactorSession)) {
+    return false;
+  }
+
+  return true;
+}
+
+async function storeTwoFactorSession(
+  req: Request,
+  options: ValidateRequestOptions = DefaultValidateRequestOptions,
+): Promise<void> {
+  const sessionKey = getSessionKey(req, options);
+  return cache.set(sessionKey, {}, options.sessionDuration);
+}
+
+async function validateRequest(
+  req: Request,
+  options: ValidateRequestOptions = DefaultValidateRequestOptions,
+): Promise<void> {
+  options = { ...DefaultValidateRequestOptions, ...options };
+
+  const remoteUser = req?.remoteUser;
```

Comment:
Unless I'm missing something, I don't think we want to allow a null request here as it would crash `userHasTwoFactorAuthEnabled` anyway

```suggestion
  if (!req.remoteUser) {
    throw new Unauthorized();
  }
  
  const remoteUser = req.remoteUser;
```